### PR TITLE
nixos/loki: skip config validation when it's impossible to validate

### PR DESCRIPTION
**Pull request by emilylange**
## Description of changes

This is a follow-up to 8d7f3c9dbde7d235f3e256f2dc13ec07e68f60f6 and ae48735c531aaa01fb083333b1ec7e538c04b99d.

Running the config validation in the build sandbox is impossible and will fail when using `cfg.configFile` or `-config.expand-env=true`.

`cfg.configFile` is a string of a path which is simply not available to the build sandbox.

Similarly, one may opt to use `cfg.configuration` with environment variables in combination with `-config.expand-env=true`.

The environment variables referenced that way are also not available in the build sandbox.

So we skip the validation when it's impossible (`cfg.configFile`) or likely impossible (`-config.expand-env=true`).

An alternative approach would be something like nixos/prometheus' `services.prometheus.checkConfig` that takes a boolean and makes toggling the config validation user-facing.

cc @mjm (Thanks for reporting!)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by emilylange** [comment]
@ofborg test loki

File: nixos/modules/services/monitoring/loki.nix [hunk 1]
```diff
@@ -97,7 +97,12 @@ in {
 
       serviceConfig = let
         conf = if cfg.configFile == null
-               then prettyJSON cfg.configuration
+               then
+                 # Config validation may fail when using extraFlag = [ "-config.expand-env=true" ].
+                 # To work around this, we simply skip it when extraFlags is not empty.
+                 if cfg.extraFlag == []
```
Comment (mjm) [inline nixos/modules/services/monitoring/loki.nix:103]: This doesn't eval, I think it should be `extraFlags`

Comment (emilylange) [reply nixos/modules/services/monitoring/loki.nix:103]: oh whoops, sorry

**Review by mjm — APPROVED** [review]

**Comment by emilylange** [comment]
(last force push fixes the `extraFlag` vs `extraFlags` typo in the comment as well)

**Pull request merged**